### ISSUE TITLE
Sample packing for pretraining/fine-tuning

### ISSUE THREAD

**Issue by alitirmizi23** [Lightning-AI/lit-gpt#620]
I was wondering if there are sample packing approaches defined somewhere for preprocessing and tokenization of datasets? I looked through different prepare_*.py, but couldn't find anything related to packing multiple sequences being packed into max_length for efficiency etc

Also, wondering how the data prep works as of now in the lit-gpt framework:

- if a document, article, instruction/output pair exceeds the max sequence length, how is it treated? What about if a doc/article/instruction-output pair falls short of max seq. length? are the remaining time steps padded or are more sequences packed until max length is achieved?

**Comment by carmocca** [comment]
cc @awaelchli, if you'd like to answer

**Comment by jwkirchenbauer** [comment]
Just chiming in, from what I understand, this is not a simple feature to implement in general.

As one current example, the axolotl finetuning harness implements efficient sample packing with correct block diagonal attention masking through [a series of monkey patches](https://github.com/OpenAccess-AI-Collective/axolotl/tree/main/src/axolotl/monkeypatch) for the underlying huggingface model definitions for a few of the very popular models like llama and mistral.

It seems like the simplicity of the lit-gpt model definition might actually make this easier to implement as a first class feature. It is relevant for efficient finetuning (the reason it's incorporated into axolotl), and general wisdom (and whispers from inside large corps) suggest that this type of block diagonal masking is standard in production training code.

I (and other collaborators) would be very interested in this feature and it would increase the attractiveness of lit-gpt's model building code as a hf alternative. Just my 2c!

**Comment by carmocca** [comment]
> if a document, article, instruction/output pair exceeds the max sequence length, how is it treated?

Depends on the data preparation, but our scripts trim it: https://github.com/Lightning-AI/lit-gpt/blob/0791c52a944f022a5cee91ed1e47288830efb72c/scripts/prepare_alpaca.py#L116-L117

> What about if a doc/article/instruction-output pair falls short of max seq. length? are the remaining time steps padded or are more sequences packed until max length is achieved?

They are packed in the pretraining scripts: https://github.com/Lightning-AI/lit-gpt/blob/0791c52a944f022a5cee91ed1e47288830efb72c/finetune/full.py#L232-L246
and padded in the fine-tuning scripts: https://github.com/Lightning-AI/lit-gpt/blob/0791c52a944f022a5cee91ed1e47288830efb72c/finetune/full.py#L232-L246

> suggest that this type of block diagonal masking is better for large scale training code.

The inconvenience is that `torch.nn.functional.scaled_dot_product_attention` will not use flash-attn if an attention mask is passed. It would be necessary to integrate this specific flavor of flash attention: https://github.com/Dao-AILab/flash-attention/issues/654 which would again require building it

**Comment by vgoklani** [comment]
https://github.com/Dao-AILab/flash-attention/issues/432#issuecomment-1698610752

**Comment by corbt** [comment]
We'd also be very interested in this feature!

**Comment by lantiga** [comment]
@carmocca let’s revive this issue
it doesn’t look like spda from PyTorch 2.3 has solved the underlying issue, if that’s the case let’s add flash attention as an optional dependency

**Comment by carmocca** [comment]
PyTorch has added support for arbitrary custom masks as long which are meant to be performant when used with `torch.compile`: https://github.com/pytorch/pytorch/pull/121845

They are also considering more generic API changes that are in discussion: https://github.com/pytorch/pytorch/issues/110681.

As of today, Tri Dao's package is the only option as far as I know.

**Comment by samsja** [comment]
> PyTorch has added support for arbitrary custom masks as long which are meant to be performant when used with `torch.compile`: [pytorch/pytorch#121845](https://github.com/pytorch/pytorch/pull/121845)
> 
> They are also considering more generic API changes that are in discussion: [pytorch/pytorch#110681](https://github.com/pytorch/pytorch/issues/110681).
> 
> As of today, Tri Dao's package is the only option as far as I know.

I think that xformers is doing it as well

**Comment by samsja** [comment]
oh I see, any chance litgpt will integrate some of this option at this point ?

By any chance do you have some benchmark comparing fa2/fa3/xformers/torch sdpa ?

**Comment by rasbt** [comment]
That's a good question. We don't have a benchmark but LitGPT already supports FlashAttention-2 via PyTorch's SDPA. The plan is to also support FlashAttention-3 (#1578)

**Comment by samsja** [comment]
> That's a good question. We don't have a benchmark but LitGPT already supports FlashAttention-2 via PyTorch's SDPA. The plan is to also support FlashAttention-3 (#1578)

unfortunately torch sdpa cannot leverage flash attention with custom masks (for context stuffing), contrary to xformers and the original flash attention implementation. Bit of a blocker. I am currently implementing it using xformers and litgpt

**Comment by ali-issa99** [comment]
Hello @samsja any updates on how we can apply data packing with masking to prevent data from different context to be used when computing attention?

**Comment by samsja** [comment]
> Hello @samsja any updates on how we can apply data packing with masking to prevent data from different context to be used when computing attention?

I end up using `flash_attn_varlen_func` from flash attention.  It's probably a good idea to try `flex_attention` package as well

**Comment by ali-issa99** [comment]
@samsja  and you combined it with the code provided by LitGPT? if yes, do u have any code example?

**Comment by samsja** [comment]
> @samsja and you combined it with the code provided by LitGPT? if yes, do u have any code example?

I am re using some litgpt code indeed, but had to change the attention layer as well as the dataloading. Unfortunalty cannot share code for now

**Comment by t-vi** [comment]
I'd be very interested in this, my preferred API would be:
- pass a new keyword arguments `batch_lengths`, a 1d tensor, that would give the lengths of each batch item,
- allow/require `idx` (and `input_pos` if present) to be 1d with the convention that the first `batch_lengths[0]` items pertain to batch item 0, the next `batch_lengths[1]` items to batch item 1 etc. (this is needed for the mask and also for the kvcache indexing).

In terms of implementation, the key points from my pov are:
- `batch_lengths` needs to be a tensor (same device as `idx`/`input_pos`),
- the "decoding" should be done in a functional interface, not in the modules,
- a first version might just loop over the batch items, but for example attention should be possible to implement via FlexAttention (though I think it might become fiddely),
- the implementation should enable to use more sophisticated implementation that avoid the loop down the road (this is why we need the functional interface and have things be a device tensor).

Happy to answer questions if anyone wants to tackle this.

**Comment by samsja** [comment]
my two cents opinion after having implemented this multiples times:

* its easier to just pack everything along the sequence dimension (basically having a 1, seq*batch_size) input ids. This allow to not cut the data in half when not needed

* flex attention is perfect for this. Since the that blockmask is the same for all transformers layers its easier to build the mask once at the beginning of the forward rather than building at each attention. So its IMO easier to just pass flex attention block mask as input rather than raw seqlens. Example of how I do it https://github.com/samsja/trainer/blob/dae7fb90a9e97aa8eb8bb11ad25b35e9b63c7e4e/src/zeroband/train.py#L150  https://github.com/samsja/trainer/blob/dae7fb90a9e97aa8eb8bb11ad25b35e9b63c7e4e/src/zeroband/models/llama/model.py#L298


other nice implementation https://github.com/KellerJordan/modded-nanogpt  https://github.com/pytorch/torchtune

**Comment by samsja** [comment]
> > [@samsja](https://github.com/samsja) and you combined it with the code provided by LitGPT? if yes, do u have any code example?
> 
> I am re using some litgpt code indeed, but had to change the attention layer as well as the dataloading. Unfortunalty cannot share code for now

the code that I was referring to that is based on litgpt is here btw https://github.com/metagene-ai/metagene-pretrain/blob/82b9e142db2c7e0a268346d53344d6c3bf223066/train/litgpt/pretrain.py#L365 using flash_attn_2
